crates/language/src/language_registry.rs

```diff
@@ -485,7 +486,16 @@ impl LanguageRegistry {
                 .and_then(|types| types.get(language_name))
                 .unwrap_or(&empty)
                 .iter()
-                .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())));
+                .any(|suffix| {
+                    path_suffixes.iter().any(|path_suffix| {
```

Comment:
Here we'd be parsing globs on each invokation of `language_for_file_internal`, which is a pretty hot function; I think a better option would be to change type of `user_file_types` to `Option<&HashMap<Arc<str>, Vec<Pattern>>>` (and the corresponding type in settings: https://github.com/zed-industries/zed/blob/main/crates/language/src/language_settings.rs#L57) and do the parsing in settings deserialization (https://github.com/zed-industries/zed/blob/main/crates/language/src/language_settings.rs#L566). That way, invalid globs will be rejected at settings load time, which should be better. 

Also, mea culpa but we already have a dependency for glob parsing (`globset`) - we can use that instead of `glob` as it should also be presumably more efficient for our use case (of multiple globs) with [GlobSet](https://docs.rs/globset/latest/globset/struct.GlobSet.html). This is my fault, as I've suggested using glob initially. Sorry!

Overall though this looks good to me; once we move glob parsing to Settings-load time, we should be good to merge :)

Reply:
I started refactor the code to accept `Option<&HashMap<Arc<str>, Vec<Pattern>>>` on `AllLanguageSettings` but I faced some issues to add the Lifetime parameter (need a lot of refactors to add it), so I add the type of `file_types` as `Option<HashMap<Arc<str>, GlobSet>>`, treated it on `impl settings::Settings for AllLanguageSettings` method like this:
```
            for (language, suffixes) in &user_settings.file_types {
                let mut builder = GlobSetBuilder::new();
                for suffix in suffixes {
                    builder.add(Glob::new(suffix)?);
                }
                file_types
                    .as_mut()
                    .unwrap()
                    .insert(language.clone(), builder.build()?);
            }
```
With this I need get the reference on https://github.com/zed-industries/zed/blob/32e642454329ac4ad469e032ff439afd465b771b/crates/language/src/language_registry.rs#L481 using `user_file_types.file_types.as_ref()` and refactor the `path_matches_custom_suffix` to be something like:
```
            let path_matches_custom_suffix = user_file_types
                .and_then(|types| types.get(language_name))
                .map_or(false, |globset| {
                    path_suffixes
                        .iter()
                        .any(|&suffix| suffix.map_or(false, |suffix| globset.is_match(suffix)))
                });
```
I'll push the changes, if have some appointments, how I said, I'm a newcomer and I little bit afraid to make some big changes on the code.

Reply:
Yeah, what you have is pretty much fine; I didn't mean to use a reference type on the `AllLanguageSettings` itself, since - as you've pointed out - we'd have to carry that lifetime around everywhere, and that would be a pain in the bum.